pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
 	}
 }
 
+func (s *server) processDelta(str stream.DeltaStream, reqCh <-chan *discovery.DeltaDiscoveryRequest, defaultTypeURL string) error {
+	// increment stream count
+	streamID := atomic.AddInt64(&s.streamCount, 1)
+
+	// unique nonce generator for req-resp pairs per xDS stream; the server
+	// ignores stale nonces. nonce is only modified within send() function.
+	var streamNonce int64
+
+	// a collection of stack alloceated watches per request type
+	var values watches
+	values.Init()
+
+	defer func() {
+		values.Cancel()
+		if s.callbacks != nil {
+			s.callbacks.OnDeltaStreamClosed(streamID)
+		}
+	}()
+
+	// sends a response by serializing to protobuf Any
+	send := func(resp cache.DeltaResponse, typeURL string) (string, error) {
+		if resp == nil {
+			return "", errors.New("missing response")
+		}
+
+		out, err := resp.GetDeltaDiscoveryResponse()
+		if err != nil {
+			return "", err
+		}
+
+		// increment nonce
+		streamNonce = streamNonce + 1
+		out.Nonce = strconv.FormatInt(streamNonce, 10)
+		if s.callbacks != nil {
+			s.callbacks.OnStreamDeltaResponse(streamID, resp.GetDeltaRequest(), out)
+		}
+
+		return out.Nonce, str.Send(out)
+	}
+
+	// setState updates the currently known state of resources in the server
+	setState := func(res cache.DeltaResponse, typeURL string) {
+		values.mu.Lock()
+		defer values.mu.Unlock()
+
+		state := values.deltaStreamStates[typeURL]
+		if state.ResourceVersions == nil {
+			state.ResourceVersions = make(map[string]string)
+		}
+		state.ResourceVersions = res.GetNextVersionMap()
+		values.deltaStreamStates[typeURL] = state
+	}
+
+	if s.callbacks != nil {
+		if err := s.callbacks.OnDeltaStreamOpen(str.Context(), streamID, defaultTypeURL); err != nil {
+			return err
+		}
+	}
+
+	// node may only be set on the first discovery request
+	var node = &core.Node{}
+	isWildcard := map[string]bool{}
+
+	for {
+		select {
+		case <-s.ctx.Done():
+			return nil
+			// config watcher can send the requested resources types in any order
+		case resp, more := <-values.deltaEndpoints:
+			if !more {
+				return status.Errorf(codes.Unavailable, "endpoints watch failed")
+			}
+			nonce, err := send(resp, resource.EndpointType)
+			if err != nil {
+				return err
+			}
+			values.deltaEndpointNonce = nonce
+			setState(resp, resource.EndpointType)
```

Comment:
Have you looked at how it would look if we tried to extract this into a helper? We don't have any type information in the per type channels etc., so we should be able to extract common functionality?

Reply:
I'm guessing you're talking about this whole block in the select statement? I have actually, I didn't like the repeated code here. I thought about creating a map to hold the nonces that way we don't have separate fields in the watches struct, so it would be a lot easier to deal with since we can pull the type info out of the `resp` to set that. I'll see what I can do here